### ISSUE TITLE
[selectors-4] Rename :focus-ring to :focus-indicator

### ISSUE THREAD

**Issue by alice** [w3c/csswg-drafts#2036]
https://drafts.csswg.org/selectors-4/#the-focusring-pseudo

- The current name causes confusion as authors tend to assume it refers to modifying the style of the default focus ring
- It is also misleading as it seems to imply that the author-provided focus indicator should be a ring, whereas in many cases author-provided focus styles are not

`:focus-indicator` does not carry this semantic baggage.

Furthermore, the phrase "focus indicator" is used in [WCAG guidelines](https://www.w3.org/TR/WCAG20-TECHS/G195.html), creating a semantic link between these two notions.

**Comment by alice** [comment]
A couple of comments on the status quo regarding focus indicators in general:

- Chrome and Firefox and Edge all set focus on click. Safari does not. 
- The CSS spec has nothing at all to say about when a focus ring *should* be drawn.
  - This is WAI - browsers should be able to determine when they wish to show a focus indicator.
  - This selector will allow styling *that* case, as opposed to `:focus` which will (and should continue to) match whenever the element is focused.
- The current behaviour for a `<button>` element in Chrome and Edge is what we wish to normalise:
  - Clicking sets focus
  - `:focus` matches, but no focus ring is drawn
- In Chrome, this is effectively a hack: there is a user agent style for `:focus` which matches and should apparently be applied, but it is hard coded not to show on click.
  - For elements without this special exception, the UA style matches and a focus ring is shown.

**Comment by alice** [comment]
http://output.jsbin.com/kijafe/1 demonstrates the difference between `<button>` default focus styling and `<div>` (or custom element) default focus styling.

**Comment by bkardell** [comment]
Just wanted to add that Alice and I have been talking about adding this to the agenda since the week before TPAC, I 100% support this. Just to add briefly: The fact that ':focus-ring' isn't necessarily a ring, nor does it necessarily have a 100% standard 'thing' that it does (some people use border, some people use backgrounds, certain things indicate just fine without) has been confusing, in actual practice, for a number of authors and testers I've encountered.  In all cases this was resolved by pointing to existing documents explaining that it is really a focus _indicator_ we're interested in.

**Comment by FremyCompany** [comment]
I have no strong opinion, but since there is a call to rename, I'll add my two cents to the discussion and propose `:focus-visible` (but do not intend to argue about it, it's just an idea for you to consider)

**Comment by AmeliaBR** [comment]
What about `:focus-needed`, to match [the Canvas 2D API `drawFocusIfNeeded` method](https://html.spec.whatwg.org/multipage/canvas.html#drawing-focus-rings-and-scrolling-paths-into-view)?

I agree with the criticisms about `:focus-ring`.

`:focus-indicator` is rather long and technical, and still sounds like a pseudoelement (styling the focus indicator) rather than a pseudoclass (the state of needing a visible focus indicator).

I don't like `:focus-shown` because (unlike with `:placeholder-shown`) this isn't about adjusting for the browser's focus rendering, it's about replacing it. (And depending on the browser and other styling, making sure it's there in the first place.)

I also like the idea that a proper focus indicator (in some form) is _needed_ when this pseudoclass is active.  In contrast to the broader `:focus` pseudoclass, where it would be more of a stylistic choice.

**Comment by FremyCompany** [comment]
`:focus-needed` is compelling

**Comment by alice** [comment]
I don't like `:focus-needed` since that seems to refer to the state of "needing focus", rather than "needing a visible focus indicator".

`:focus-visible` would be a good alternative IMO.

**Comment by bkardell** [comment]
Matching existing terms in other existing recommendations, documentation and training seems like a big win to me.  I don't think that saving 2 characters is a great reason to choose `:focus-visible` over `:focus-indicator` if it also comes at the cost of losing that benefit.  That said, I will agree that even `:focus-visible` would be better than `:focus-indicator`.  What about :focus-pocus? (yes, I am kidding).

**Comment by tabatkins** [comment]
In previous discussions with Alice, I objected to :focus-indicator, because it's just a more generic version of the existing term and thus presumably falls to the same confusion about "is this styling the focus ring itself?" - you can still ask "is this styling the focus indicator itself?"

Switching to a state-based name, like :focus-visible, addresses that concern, so 👍 . (It's also slightly simpler to spell, so yay.)

**Comment by alice** [comment]
It sounds like there is grudging consensus on `:focus-visible`.

Any objections? I'm happy to send a patch to the spec if not.

cc @robdodson

**Comment by bkardell** [comment]
I agree it is better than `:focus-ring`. If the rest of group doesn't buy the rationale provided for why `:focus-indicator` is actually a term that is both widely used and has adequately cleared up the confusion in IRL use, then I'm fine with `:focus-visible`.

**Comment by bkardell** [comment]
@alice To be clear: I am 100% fine with the rationale of the group if the choice is `:focus-visible` :) One I have seen work and matches existing docs, the other is speculative but seems very good.

**Comment by frivoal** [comment]
+1 for focus-visible, with focus-shown and :focus-indicated as acceptable second choices.
